pymatviz/heatmap.py

```diff
@@ -0,0 +1,91 @@
+"""Heatmap plotter."""
+
+
+from __future__ import annotations
+
+from typing import Any
+
+import matplotlib.pyplot as plt
+import pandas as pd
+from matplotlib.colors import Colormap
+
+
+def heatmap(
+    df: pd.DataFrame,
+    cmap: Colormap,
+    add_cbar: bool = True,
+    cbar_label: str = "Colorbar label",
+    ax: plt.Axes = None,
+    x_label: str = "X label",
+    y_label: str = "Y label",
+    **kwargs: Any,
+) -> plt.Axes:
+    """Create a heatmap from a 2D Pandas DataFrame.
+
+    Parameters:
+    - df (pd.DataFrame): The input DataFrame containing the data to visualize.
+        The index is used as tick labels.
+    - cmap (Colormap): Matplotlib colormap to be used for coloring the heatmap.
+    - add_cbar (bool, optional): Whether to add a colorbar. Default is True.
+    - cbar_label (str, optional): Label for the colorbar.
+        Default is "Colorbar label".
+    - ax (plt.Axes, optional): Matplotlib Axes on which to plot the heatmap.
+        If not provided, the current Axes is used.
+    - x_label (str, optional): Label for the x-axis. Default is "X label".
+    - y_label (str, optional): Label for the y-axis. Default is "Y label".
+
+    Returns:
+    - plt.Axes: Matplotlib Axes containing the heatmap.
+
+    Raises:
+    - AssertionError: If input types or values are not as expected.
+
+    Example:
+    ```python
+    # Create a sample DataFrame
+    data_df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': [7, 8, 9]})
+
+    # Plot the heatmap
+    fig, ax = plt.subplots()
+    heatmap(data_df, cmap, ax=ax, x_label='X-Axis', y_label='Y-Axis')
+    ax.set_title('Heatmap Title')
+
+    plt.show()
+    ```
+    """
+    # Type and value checks
+    if not isinstance(df, pd.DataFrame):
+        raise TypeError("Heatmap plotter expect a pd.DataFrame as input.")
+    if not isinstance(cmap, Colormap):
+        raise TypeError(
+            f"Heatmap plotter expect a Colormap instance, got {type(cmap)}."
+        )
+    if not isinstance(add_cbar, bool):
+        raise TypeError("add_cbar should be bool.")
+
+    if df.ndim != 2:
+        raise ValueError("Heatmap plotter expect a 2D DataFrame.")
+
+    # Initialize ax for plotting
+    ax = ax or plt.gca()
+
+    # Generate heatmap
+    im = plt.imshow(df, cmap)
+
+    # Set x/y axes tick labels
+    ax.set_xticks(range(len(df.columns)))
+    ax.set_xticklabels(df.columns, rotation="vertical")
+
+    ax.set_yticks(range(len(df.index)))
+    ax.set_yticklabels(df.index)
+
+    # Set x/y axes labels
+    ax.set_xlabel(x_label)
+    ax.set_ylabel(y_label)
+
+    # Add colorbar
+    if add_cbar:
+        cbar = ax.figure.colorbar(im, ax=ax)
+        cbar.set_label(cbar_label)
+
+    return ax
```

Comment:
`heatmap()` currently looks a bit like a simple wrapper around `plt.imshow()`. i think a good overarching design principle for functions is how much effort does it save to use the function. to do more in this function, we could make it a bit more task-specific. in the context of materials science, this might be naming the function `element_heatmap()` and defaulting the x/y axis tick labels to element symbols (H, He, Li, Be, ...) so that people can pass np.array/pd.DataFrame without needing to have element symbols as column/row indices

Reply:
Hi @janosh , thanks for reviewing and these suggestions. Yes it's only after I begin to work on this heatmap function that I realized it's simply a wrappler of `imshow()`. 

For me personally the tricky part for this heatmap plot is actually the manipulation (combine/truncate) of colormaps in an intuitive manner (I would talk more about this in the following specific parts). I might go ahead and separate the column/row labels from the single `pd.DataFrame` to make altering those easier.

However I personally want to make this plotter more general and people could pass in any 2D data array? So I didn't introduce too much task specific parameters...

Another thing might be to add an "interpolation ()" to make generation of the below "smooth" heatmap easier, and to allow labelling data points (but I guess that could be also made into a "decorator" like function to wrap the original function to add some points).

![A smooth heatmap](https://github.com/janosh/pymatviz/assets/80093591/5678252c-ea80-4a70-bbdf-ff00309d1ecf)